metatensor-torch/src/tensor.cpp

```diff
@@ -52,7 +91,29 @@ static std::vector<metatensor::TensorBlock> blocks_from_torch(const std::vector<
 
 TensorMapHolder::TensorMapHolder(TorchLabels keys, const std::vector<TorchTensorBlock>& blocks):
     tensor_(keys->as_metatensor(), blocks_from_torch(blocks))
-{}
+{
+    if (blocks.size() == 0) return;  // nothing to check
+
+    const auto device = keys->values().device();
+    // REVIEWME: Is this necessary or not?
+    // We could proceed like in the other constructor and only check the devices of the blocks.
+    
+    
+    const auto dtype = blocks[0]->values().dtype();
+
+    for (const auto& block : blocks) {
+        if (block->values().device() != device) {
+            C10_THROW_ERROR(TypeError,
+                "cannot create TensorMap: keys and blocks must be on the same device"
+            );
+        }
+        if (block->values().dtype() != dtype) {
+            C10_THROW_ERROR(TypeError,
+                "cannot create TensorMap: all blocks must have the same dtype"
+            );
```

Comment:
same comments as above: this should be a TypeError, and the error message should contain some information about the encoutered device/dtype.